### PR TITLE
Payments API

### PR DESCRIPTION
**Pull request by MarcoBuster**

### CODE REVIEW

File: botogram/hooks.py [hunk 1]
```diff
@@ -230,6 +230,24 @@ def _call(self, bot, update):
                          message=message)
 
 
+class ShippingQueryHook(Hook):
+    """Underlying hook for @bot.shipping_query"""
+
+    def _call(self, bot, update):
+        shipping_query = update.shipping_query
+        return bot._call(self.func, self.component_id,
+                         shipping_query=shipping_query)
```
Comment (pietroalbini) [inline botogram/hooks.py]: Maybe `shipping_query` can be called only `query`:

```python
@bot.shipping_query
def return_shipping(bot, query):
    query.decline("Please try tomorrow.")
```

File: botogram/hooks.py [hunk 2]
```diff
@@ -230,6 +230,24 @@ def _call(self, bot, update):
...
+
+
+class PreCheckoutQueryHook(Hook):
+    """Underlying hook for @bot.pre_checkout_query"""
+
+    def _call(self, bot, update):
+        pre_checkout_query = update.pre_checkout_query
+        return bot._call(self.func, self.component_id,
+                         pre_checkout_query=pre_checkout_query)
```
Comment (pietroalbini) [inline botogram/hooks.py]: Maybe pre_checkout_query can be called only query:

```python
@bot.pre_checkout
def out_of_stock_sorry(bot, query):
    query.decline("I don't like you.")
```

File: botogram/bot.py [hunk 3]
```diff
@@ -158,6 +163,16 @@ def channel_post_edited(self, func):
         self._main_component.add_channel_post_edited_hook(func)
         return func
 
+    def shipping_query(self, func):
+        """Add a shipping query hook"""
+        self._main_component.add_shipping_query_hook(func)
+        return func
+
+    def pre_checkout_query(self, func):
```
Comment (pietroalbini) [inline botogram/bot.py]: I would remove the `_query` from the name of the method, something like

```python
@bot.pre_checkout
def accept_blindly(query):
    query.decline()
```

File: botogram/bot.py [hunk 4]
```diff
@@ -158,6 +163,16 @@ def channel_post_edited(self, func):
         self._main_component.add_channel_post_edited_hook(func)
         return func
 
+    def shipping_query(self, func):
```
Comment (pietroalbini) [inline botogram/bot.py]: I don't like `shipping_query`, but I can't think about a replacement right now.

File: botogram/components.py [hunk 5]
```diff
@@ -176,6 +178,22 @@ def add_channel_post_edited_hook(self, func):
...
 
+    def add_shipping_query_hook(self, func):
+        """Add a shipping query hook"""
+        if not callable(func):
+            raise ValueError("A shipping query hook must be callable")
+
+        hook = hooks.ShippingQueryHook(func, self)
+        self.__shipping_query_hooks.append(hook)
+
+    def add_pre_checkout_query_hook(self, func):
```
Comment (pietroalbini) [inline botogram/components.py]: Remove the `_query` here

File: botogram/components.py [hunk 6]
```diff
@@ -176,6 +178,22 @@ def add_channel_post_edited_hook(self, func):
         hook = hooks.EditedChannelPostHook(func, self)
         self.__channel_post_edited_hooks.append(hook)
 
+    def add_shipping_query_hook(self, func):
```
Comment (pietroalbini) [inline botogram/components.py]: Let's rename this as the decorator, when we decide a name.

File: botogram/payments.py [hunk 7]
```diff
@@ -0,0 +1,76 @@
...
+
+    def add(self, title, prices):
+        """Add a options to options"""
+        telegram_prices = []
+        for price in prices:
+            telegram_prices.append({"label": price, "amount": prices[price]})
+
+        self._options.append({"title": title, "prices": telegram_prices})
+
+    def _to_json(self):
```
Comment (pietroalbini) [inline botogram/payments.py]: This method should be named `serialize`.

File: docs/api/telegram.rst [hunk 8]
```diff
@@ -372,6 +374,28 @@ about its business.
...
+      :param int photo_size: Photo size
+      :param int photo_width: Photo width
+      :param int photo_height: Photo height
+      :param bool need_name: Pass True, if you require the user's full name to complete the order
+      :param bool need_phone_number: Pass True, if you require the user's phone number to complete the order
+      :param bool need_email: Pass True, if you require the user's email to complete the order
+      :param bool need_shipping_address: Pass True, if you require the user's shipping address to complete the order
+      :param bool is_flexible: Pass True, if the final price depends on the shipping method
+      :param bool notify: If you want to trigger a notification on the client
+
```
Comment (pietroalbini) [inline docs/api/telegram.rst]: You forgot the versionadded. Also, why you documented a thing you know will change? :P

File: docs/api/telegram.rst [hunk 9]
```diff
@@ -2077,6 +2117,93 @@ about its business.
...
+             shipping_options.add("Amazon 2 days", {"Spedition": 1000})
+             shipping_options.add("Amazon 5 days", {"Spedition": 500})
+             shipping_query.reply(shipping_options)
+
+
+   .. py:method:: decline(error_message)
+
+      Decline the shipping address with an error message.
+
+   .. versionadded:: 0.4
```
Comment (pietroalbini) [inline docs/api/telegram.rst]: This versionadded should be added after the description of the class.

File: docs/api/telegram.rst [hunk 10]
```diff
@@ -2077,6 +2117,93 @@ about its business.
...
+
+   .. py:method:: accept()
+
+      Accept the pre checkout and continue the payment.
+
+   .. py:method:: decline(error_message)
+
+      Decline the pre checkout and abort the payment with an error message.
+
+   .. versionadded:: 0.4
```
Comment (pietroalbini) [inline docs/api/telegram.rst]: This versionadded should be added after the description of the class.

File: docs/api/telegram.rst [hunk 11]
```diff
@@ -2077,6 +2117,93 @@ about its business.
...
+
+   .. py:method:: decline(error_message)
+
+      Decline the shipping address with an error message.
+
+   .. versionadded:: 0.4
+
+
+.. py:class:: botogram.PreCheckoutQuery
+
```
Comment (pietroalbini) [inline docs/api/telegram.rst]: Forgot the class description.

File: botogram/bot.py (hunk 3 above)
Comment (MarcoBuster) [inline botogram/bot.py]: Ok!

File: botogram/components.py (hunk 6 above)
Comment (MarcoBuster) [inline botogram/components.py]: Ok!

File: botogram/components.py (hunk 5 above)
Comment (MarcoBuster) [inline botogram/components.py]: Ok!

File: botogram/hooks.py (hunk 1 above)
Comment (MarcoBuster) [inline botogram/hooks.py]: Ok!

File: botogram/hooks.py (hunk 2 above)
Comment (MarcoBuster) [inline botogram/hooks.py]: Ok!

File: botogram/payments.py (hunk 7 above)
Comment (MarcoBuster) [inline botogram/payments.py]: Ok!

File: docs/api/telegram.rst (hunk 9 above)
Comment (MarcoBuster) [inline docs/api/telegram.rst]: Ok!

File: docs/api/telegram.rst (hunk 10 above)
Comment (MarcoBuster) [inline docs/api/telegram.rst]: Ops!

File: docs/api/telegram.rst (hunk 11 above)
Comment (MarcoBuster) [inline docs/api/telegram.rst]: Ops!

**Comment by pietroalbini** [comment]
Thinking a bit more about this, I don't think adding wrapper objects for a nearly unused feature of the API is worth. It's just increased maintenance for the project, and because payments are practically limited only to phisical things (iOS prevents digital purchases) the API is also mostly useless.

If someone really wants to use the feature, there is still the ability to add custom objects and processors, so not having native support is not a big deal.

If someone has compelling arguments about this I'm willing to reopen the PR :)

**Pull request closed without merging**